File: Groupe2/TP2/src/etudiant.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>  // Nécessaire pour utiliser strcpy()

struct Etudiant {
    char nom[100];
    char prenom[100];
    char adresse[100];
    int *notes;
    int nombre_notes;
};

void afficher_etudiant(struct Etudiant etudiant) {
    printf("Nom : %s\n", etudiant.nom);
    printf("Prenom : %s\n", etudiant.prenom);
    printf("Adresse : %s\n", etudiant.adresse);
    printf("Notes : ");
    for (int i = 0; i < etudiant.nombre_notes; i++) {
        printf("%d ", etudiant.notes[i]);
    }
    printf("\n");
}

int main() {
    struct Etudiant etudiants[5];  // Tableau de 5 étudiants

    for (int i = 0; i < 5; i++) {
        printf("Saisir les informations pour l'etudiant.e %d\n", i + 1);
        printf("=========================================\n");
        char buffer[100];  // Buffer temporaire pour stocker les entrées de l'utilisateur

        // Saisie des informations avec strcpy
        printf("Saisir le nom de l'etudiant.e : ");
        scanf("%s", buffer);
        strcpy(etudiants[i].nom, buffer);
        printf("Saisir le prenom de l'etudiant.e : ");
        scanf("%s", buffer);
        strcpy(etudiants[i].prenom, buffer);
        printf("Saisir l'adresse de l'etudiant.e : ");
        scanf("%s", buffer);
        strcpy(etudiants[i].adresse, buffer);

        // Saisie du nombre de notes pour allouer la mémoire dynamiquement
        printf("Combien de notes voulez-vous saisir ? ");
        scanf("%d", &etudiants[i].nombre_notes);

        // Vérification que le nombre de notes est valide
        if (etudiants[i].nombre_notes <= 0) {
            printf("Nombre de notes invalide.\n");
            return 1;
        }

        // Allocation dynamique de la mémoire pour les notes
        etudiants[i].notes = (int*)malloc(etudiants[i].nombre_notes * sizeof(int));
        if (etudiants[i].notes == NULL) {
            printf("Erreur d'allocation de mémoire\n");
            return 1;
        }

        // Saisie des notes de l'étudiant et stockage dans le tableau dynamique
        for (int j = 0; j < etudiants[i].nombre_notes; j++) {
            printf("Note %d: ", j + 1);
            scanf("%d", &etudiants[i].notes[j]);
        }
    }

    // Affichage des informations des étudiants avec la fonction afficher_etudiant
    for (int i = 0; i < (sizeof(etudiants) / sizeof(etudiants[0])); i++) {
        printf("Informations pour l'etudiant.e %d\n", i + 1);
        printf("=========================================\n");
        afficher_etudiant(etudiants[i]);

        free(etudiants[i].notes);  // Libérer la mémoire allouée dynamiquement pour les notes
        etudiants[i].notes = NULL;  // Bonne pratique : on annule le pointeur après libération
    }

    return 0;
}

File: Groupe2/TP2/src/etudiant.py
class Etudiant:
    """
    Classe représentant un.e étudiant.e
    """

    def __init__(
            self,
            nom: str,
            prenom: str,
            adresse: str,
            notes: list
    ):
        self.nom = nom
        self.prenom = prenom
        self.adresse = adresse
        self.notes = notes

    def __str__(self):
        return f"Nom : {self.nom} \nPrénom : {self.prenom} \nAdresse : {self.adresse} \nNotes : {', '.join(map(str, self.notes))}"


if __name__ == "__main__":
    # Créer une liste d'étudiants
    etudiants: list[Etudiant] = []

    # Demander à l'utilisateur le nombre d'étudiants à ajouter
    nb_etudiants = int(input("Combien d'étudiant.e.s voulez-vous ajouter ? "))

    # Ajouter chaque étudiant.e à la liste
    for i in range(nb_etudiants):
        print(f"========== Etudiant.e n°{i + 1} ==========")

        # Demander les informations de l'étudiant.e
        nom = input("Nom : ")
        prenom = input("Prénom : ")
        adresse = input("Adresse : ")

        # Demander le nombre de notes à ajouter
        nb_notes = int(input("Combien de notes voulez-vous ajouter ? "))
        notes = []

        # Ajouter chaque note à la liste
        for j in range(nb_notes):
            note = float(input(f"Note n°{j + 1} : "))
            notes.append(note)

        # Ajouter l'étudiant.e à la liste
        etudiants.append(Etudiant(nom, prenom, adresse, notes))

    # Afficher la liste des étudiants
    print("\n\nListe des étudiants :\n")
    for etudiant in etudiants:
        print("==========")
        print(etudiant)
        print("==========\n")
